wallet, logging: Replace WalletLogPrintf() with LogInfo()

ryanofsky opened the pull request:
Make new logging macros `LogDebug()`, `LogTrace()`, `LogInfo()`, `LogWarning()`, and `LogError()` compatible with wallet code and drop custom `WalletLogPrintf()` function. The new logging macros introduced in #28318 weren't previously useful in wallet code because wallet code prefixes most log messages with the wallet names to be be able to distinguish log output from multiple wallets. Fix this by introducing a new `WalletLogContext` class inheriting from `util::log::Context` which can include the wallet name in log messages.

This change makes `-logsourcelocations` usable with wallet code because it logs actual source locations instead of the `WalletLogPrintf()` location.

---

<!-- begin based-on -->
**This is based on #34778 + #29256.** The non-base commits are:

- [`1b8e2f75ce0` wallet, logging: Replace WalletLogPrintf() with LogInfo()](https://github.com/bitcoin/bitcoin/pull/30343/commits/1b8e2f75ce06d6ea9934b221e2b7526913e81027)
- [`1ca4092fd86` wallet, logging: Switch LogInfo to LogWarning/LogError](https://github.com/bitcoin/bitcoin/pull/30343/commits/1ca4092fd86deb184bf2f0822c56266ca2aac521)<!-- end -->

ryanofsky committed:
doc: Add documentation about log levels and macros

ryanofsky committed:
log test: Add test for all accepted logging arguments

Add new logging test to call macros with all allowed combinations of macro
arguments.

The new test replaces a less comprehensive test that doesn't cover log
statements without format arguments. It's also moved to the top of the test
suite because it's a good illustration of what typical log prints look like,
what logging calls are allowed and disallowed, and what the resulting log
output looks like.

ryanofsky committed:
log refactor: Allow log macros to accept context arguments

Allow LogDebug(), LogTrace(), LogInfo(), LogWarning(), and LogError() macros to
accept context arguments to provide more information in log messages and more
control over logging to callers.

This functionality is used in followup PRs:

- https://github.com/bitcoin/bitcoin/pull/30342 - To let libbitcoinkernel send
  output to specfic `BCLog::Logger` instances instead of a global instance, so
  output can be disambiguated and applications can have more control over
  logging.

- https://github.com/bitcoin/bitcoin/pull/30343 - To replace custom
  `WalletLogPrintf` calls with standard logging calls that automatically include
  wallet names and don't log everything at info level.

This commit does not change behavior of current log prints or require them to
be updated. It includes tests and documentation covering the new functionality.

Co-Authored-By: Hodlinator <172445034+hodlinator@users.noreply.github.com>
Co-Authored-By: stickies-v <stickies-v@protonmail.com>

hodlinator reviewed (commented):
Went through this and the parent PR together. It was helpful to see how that other PR's changes were used in this one.

ryanofsky committed:
wallet, logging: Switch LogInfo to LogWarning/LogError

Switch to LogWarning / LogError in obvious places where wallet code is clearly outputting warnings and errors.

Suggested by hodlinator in
https://github.com/bitcoin/bitcoin/pull/30343#discussion_r1655406670
https://github.com/bitcoin/bitcoin/pull/30343#discussion_r1655407301
https://github.com/bitcoin/bitcoin/pull/30343#discussion_r1655436010

ryanofsky reviewed (commented):
Thanks for the review! The suggestions are a big improvement and make the logging code more succinct and clean.

Rebased 5b30f17c22ade44d301aed5051a398fa38251ef7 -> 5a629a60c4739b94383596f9d040502b7b2c8b69 ([`pr/gwlog.1`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.1) -> [`pr/gwlog.2`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.1-rebase..pr/gwlog.2)) on top of the updated parent PR implementing review suggestions.

ryanofsky commented:
Updated 5a629a60c4739b94383596f9d040502b7b2c8b69 -> db8c6fc2ca5e1ad99aeda741e3b453f1f7bf7955 ([`pr/gwlog.2`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.2) -> [`pr/gwlog.3`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.3), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.2..pr/gwlog.3)) fixing initialization order bug in last push

ryanofsky commented on src/wallet/wallet.cpp:
```diff
@@ -3866,13 +3868,13 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
 
     if (!IsWalletFlagSet(WALLET_FLAG_DESCRIPTORS)) {
-        WalletLogPrintf("Cannot add WalletDescriptor to a non-descriptor wallet\n");
+        LogInfo(m_log, "Cannot add WalletDescriptor to a non-descriptor wallet\n");
```
> Are you sure that `LogError` is the right level? It should only be used for fatal errors where the node needs to shut down, according to the docs. This seems like a warning here, as suggested above.

Oh, that is interesting. The descriptions of those log levels in https://github.com/bitcoin/bitcoin/blob/b27afb7fb774809c9c075d56e44657e0b607a00c/doc/developer-notes.md are much more severe than I would have expected. Based on those descriptions I would expect LogError to be called LogFatal and LogWarning to be called LogCritical. I would have expected LogError to be used for normal errors, and LogWarning to be used for normal warnings.

Maybe the naming would be worth following up on, but I don't think it would be good to do in this PR. It actually seems like LogInfo is the right log level to use here, not LogWarning. Or do you disagree? Should I just drop the last two commits in this PR?

ryanofsky committed:
log refactor: Ensure categories are not logged at info and higher levels

Previously this used to be possible through the LogPrintLevel call but now that
call is removed, this change is just an internal refactoring and has no visible
effect except in tests.

ryanofsky commented:
Updated and rebased db8c6fc2ca5e1ad99aeda741e3b453f1f7bf7955 -> 29eaec25af32e543df3cff091938c3cb8f7cce17 ([`pr/gwlog.3`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.3) -> [`pr/gwlog.4`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.3-rebase..pr/gwlog.4)) with suggestion to use `constexpr`
Rebased 29eaec25af32e543df3cff091938c3cb8f7cce17 -> 0260cd5ccc3120102c001b3ab6c5f29dcf184b58 ([`pr/gwlog.4`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.4) -> [`pr/gwlog.5`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.4-rebase..pr/gwlog.5)) due to conflict with #30406
Rebased 0260cd5ccc3120102c001b3ab6c5f29dcf184b58 -> 0d49ada2b42e5b48b21bb7cb854851f09641abbb ([`pr/gwlog.5`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.5) -> [`pr/gwlog.6`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.5-rebase..pr/gwlog.6)) on top of updated base pr/bclog.20
Rebased 0d49ada2b42e5b48b21bb7cb854851f09641abbb -> f6b93a976b0127b55705dc33caf986a3344e12a4 ([`pr/gwlog.6`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.6) -> [`pr/gwlog.7`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.6-rebase..pr/gwlog.7)) due to conflict with #30485
Rebased f6b93a976b0127b55705dc33caf986a3344e12a4 -> 4c8fe044d3241664151b3727e31ac2eaa741b717 ([`pr/gwlog.7`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.7) -> [`pr/gwlog.8`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.7-rebase..pr/gwlog.8)) due to conflicts with #26619, #30659, and other PRs
Rebased 4c8fe044d3241664151b3727e31ac2eaa741b717 -> e8c6095c75dead138e79a18146ce865e7fe6529a ([`pr/gwlog.8`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.8) -> [`pr/gwlog.9`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.8-rebase..pr/gwlog.9)) due to conflicts with #31061
<!-- begin push-10 -->
Rebased e8c6095c75dead138e79a18146ce865e7fe6529a -> af6f854df0d0dd72b05cda4d2e1ed0d1c2c00738 ([`pr/gwlog.9`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.9) -> [`pr/gwlog.10`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.9-rebase..pr/gwlog.10))<!-- end -->
<!-- begin push-11 -->
Rebased af6f854df0d0dd72b05cda4d2e1ed0d1c2c00738 -> 2e4d0c3b51e2e8fbf913f28361d1a0a8d1f3e166 ([`pr/gwlog.10`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.10) -> [`pr/gwlog.11`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.11), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.10-rebase..pr/gwlog.11))<!-- end -->
<!-- begin push-12 -->
Rebased 2e4d0c3b51e2e8fbf913f28361d1a0a8d1f3e166 -> 186b721e6107a0606be6c55edbdd8ec5d5fabd2b ([`pr/gwlog.11`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.11) -> [`pr/gwlog.12`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.12), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.11-rebase..pr/gwlog.12))<!-- end -->
<!-- begin push-13 -->
Rebased 186b721e6107a0606be6c55edbdd8ec5d5fabd2b -> df507121593d4a1cd58c80e5e63735c115c728c5 ([`pr/gwlog.12`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.12) -> [`pr/gwlog.13`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.13), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.12-rebase..pr/gwlog.13))<!-- end --> on updated base PR to fix CI errors (https://github.com/bitcoin/bitcoin/actions/runs/18514132530?pr=30343), also fixing wallet name formatting causing wallet_tests/CreateWallet to fail
<!-- begin push-14 -->
Rebased df507121593d4a1cd58c80e5e63735c115c728c5 -> e84bd31c04751e968a0936b93d6cc7e639ac4409 ([`pr/gwlog.13`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.13) -> [`pr/gwlog.14`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.14), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.13-rebase..pr/gwlog.14))<!-- end -->
<!-- begin push-15 -->
Rebased e84bd31c04751e968a0936b93d6cc7e639ac4409 -> 75042af2053485badea67b1ff9b38bc820f3f24c ([`pr/gwlog.14`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.14) -> [`pr/gwlog.15`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.15), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.14-rebase..pr/gwlog.15))<!-- end -->
<!-- begin push-16 -->
Rebased 75042af2053485badea67b1ff9b38bc820f3f24c -> 74bb399dfa1f0823f26fc3bfa7917f5f4fa36668 ([`pr/gwlog.15`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.15) -> [`pr/gwlog.16`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.16), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.15-rebase..pr/gwlog.16))<!-- end -->
<!-- begin push-17 -->
Rebased 74bb399dfa1f0823f26fc3bfa7917f5f4fa36668 -> a08c9bc808ef19accc90335fb115f791a66e0158 ([`pr/gwlog.16`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.16) -> [`pr/gwlog.17`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.17), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.16-rebase..pr/gwlog.17))<!-- end --> on top of #29256 pr/bclog.34
<!-- begin push-18 -->
Rebased a08c9bc808ef19accc90335fb115f791a66e0158 -> aac55f94938aa7ac727b840a9766adb848e998ab ([`pr/gwlog.17`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.17) -> [`pr/gwlog.18`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.18), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.17-rebase..pr/gwlog.18))<!-- end --> on top of #29256 pr/bclog.36
<!-- begin push-19 -->
Rebased aac55f94938aa7ac727b840a9766adb848e998ab -> 7fc95e7cbb3ee3cae14df02602ac51a3d910945c ([`pr/gwlog.18`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.18) -> [`pr/gwlog.19`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.19), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.18-rebase..pr/gwlog.19))<!-- end --> on top of #29256 pr/bclog.37 with review suggestions
<!-- begin push-20 -->
Rebased 7fc95e7cbb3ee3cae14df02602ac51a3d910945c -> 524c5b8ef547700e4b5a234bf6bebd4d775d98e5 ([`pr/gwlog.19`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.19) -> [`pr/gwlog.20`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.20), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.19-rebase..pr/gwlog.20))<!-- end --> on top of #29256 pr/bclog.39
<!-- begin push-21 -->
Rebased 524c5b8ef547700e4b5a234bf6bebd4d775d98e5 -> 45ecab65d4434aebafa23195c3571b833faf47ca ([`pr/gwlog.20`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.20) -> [`pr/gwlog.21`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.21), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.20-rebase..pr/gwlog.21))<!-- end --> on top of #29256 pr/bclog.40
<!-- begin push-22 -->
Rebased 45ecab65d4434aebafa23195c3571b833faf47ca -> 304e4107d6f4371ee9df8ffe2f7b1a26c01f8967 ([`pr/gwlog.21`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.21) -> [`pr/gwlog.22`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.22), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.21-rebase..pr/gwlog.22))<!-- end --> on top of #29256 pr/bclog.41
<!-- begin push-23 -->
Rebased 304e4107d6f4371ee9df8ffe2f7b1a26c01f8967 -> 85b9cc0a90f69ff8b9675c6cb383c14f05629f7a ([`pr/gwlog.22`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.22) -> [`pr/gwlog.23`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.23), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.22-rebase..pr/gwlog.23))<!-- end --> on top of #29256 pr/bclog.42
<!-- begin push-24 -->
Rebased 85b9cc0a90f69ff8b9675c6cb383c14f05629f7a -> 1ca4092fd86deb184bf2f0822c56266ca2aac521 ([`pr/gwlog.23`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.23) -> [`pr/gwlog.24`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.24), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.23-rebase..pr/gwlog.24))<!-- end --> on top of #29256 pr/bclog.42 due to conflict with #35512
<!-- begin push-25 -->
Rebased 1ca4092fd86deb184bf2f0822c56266ca2aac521 -> d8caedeb36b8cd44a9b104d50d7606c27bdf97a9 ([`pr/gwlog.24`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.24) -> [`pr/gwlog.25`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.25), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.24-rebase..pr/gwlog.25))<!-- end --> on top of #29256 pr/bclog.42 due to conflict with #35955
<!-- begin push-26 -->
Rebased d8caedeb36b8cd44a9b104d50d7606c27bdf97a9 -> 956e6b50428f72ae902470070eb058d516dd2d14 ([`pr/gwlog.25`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.25) -> [`pr/gwlog.26`](https://github.com/ryanofsky/bitcoin/commits/pr/gwlog.26), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/gwlog.25-rebase..pr/gwlog.26))<!-- end --> due to conflicts with #34075

ryanofsky committed:
wallet, logging: Replace WalletLogPrintf() with LogInfo()

Make new logging macros LogDebug(), LogTrace(), LogInfo(), LogWarning(), and
LogError() compatible with wallet code and drop custom WalletLogPrintf()
function. The new logging macros introduced in #28318 weren't previously useful
in wallet code because wallet code prefixes most log messages with the wallet
names to be be able to distinguish log output from multiple wallets. Fix this
by introducing a new WalletLogContext class inheriting from BCLog::Context
which can include the wallet name in log messages

ryanofsky committed:
log refactor: Add support for custom log contexts

Custom log contexts allow overridding log formatting and adding metadata, such
as request ids or wallet names to log messages, while still using standard
macros for logging. This is used to replace WalletLogPrintf() functions with
LogInfo() calls in followup PR #30343.

stickies-v committed:
log test: verify log argument evaluation semantics

Test that LogInfo/LogWarning/LogError always evaluate their arguments
even when logging is disabled.

ajtowns pointed out this behavior was important and could affect non-logging
code if changed in
https://github.com/bitcoin/bitcoin/pull/34374#discussion_r2734793117

Co-authored-by: l0rinc <pap.lorinc@gmail.com>

adyshimony reviewed (commented):
Reviewed and tested aac55f94938a.

I found an issue, at log.h [LogPrint_](https://github.com/bitcoin/bitcoin/pull/30343/changes#diff-c75b427ca21222164fa8e95d82f29fefa6b60044d0510fe85f4e06e57cdc154bR160) , rvalue source arguments can be bound by reference and used after their lifetime ends. I tried to solve this in a few ways, but I think the most elegant simple one is to block rvalues at compile time by deleting the overload. I failed to find any rvalue usage that exists today. Solution suggested in inline comments.

I used the following test to confirm the lifetime issue (for demo of the issue only; we don't need it if we block rvalues at compile time):


```
struct LifetimeCheckedSource : public util::log::Source {
    LifetimeCheckedSource(int& counter, BCLog::Logger& logger, bool& alive)
        : Source{BCLog::VALIDATION, &logger}, m_counter{counter}, m_alive{alive}
    {
        m_alive = true;
    }
    ~LifetimeCheckedSource()
    {
        m_alive = false;
    }
    template <typename... Args>
    std::string Format(util::ConstevalFormatString<sizeof...(Args)> fmt, const Args&... args) const
    {
        CHECK_NONFATAL(m_alive);
        return tfm::format("Custom #%d %s", ++m_counter, tfm::format(fmt, args...));
    }
    int& m_counter;
    bool& m_alive;
};

BOOST_AUTO_TEST_CASE(logging_temporary_source_lifetime)
{
    BCLog::Logger logger;
    logger.m_log_timestamps = false;
    logger.EnableCategory(BCLog::LogFlags::ALL);
    logger.SetLogLevel(BCLog::Level::Debug);
    logger.StartLogging();

    std::vector<std::string> messages;
    logger.PushBackCallback([&](const std::string& s) { messages.push_back(s); });

    int counter{0};
    bool alive{false};
    auto make_source = [&counter, &logger, &alive]() {
        return LifetimeCheckedSource{counter, logger, alive};
    };
    LogDebug(make_source(), "temporary %s", "source");

    BOOST_REQUIRE_EQUAL(messages.size(), 1U);
    BOOST_CHECK_EQUAL(messages[0], "[validation] Custom #1 temporary source\n");
    BOOST_CHECK_EQUAL(counter, 1);
}

```

Run build/bin/test_bitcoin --run_test=logging_tests/logging_temporary_source_lifetime to see it fail.

adyshimony commented on src/wallet/wallet.cpp:
```diff
@@ -3782,7 +3785,7 @@ util::Result<std::reference_wrapper<DescriptorScriptPubKeyMan>> CWallet::AddWall
     auto spk_man = GetDescriptorScriptPubKeyMan(desc);
     if (spk_man) {
-        WalletLogPrintf("Update existing descriptor: %s\n", desc.descriptor->ToString());
+        LogError(m_log, "Update existing descriptor: %s\n", desc.descriptor->ToString());
```
Isn't that a happy flow? Maybe we can use LogInfo here?

ryanofsky commented on src/util/log.h:
re: https://github.com/bitcoin/bitcoin/pull/30343#discussion_r2868276474

> `#include <type_traits>`

Thanks! Added this

ryanofsky reviewed (commented):
Thanks for the reviews @adyshimony! Note that most of your comments apply to #29256 which this is based on (#29256 is subset of this PR), so you may be interested in reviewing that directly. Some of your comments are now addressed in recent updates to both PRs.

re: https://github.com/bitcoin/bitcoin/pull/30343#pullrequestreview-3871171023

> I found an issue, at log.h LogPrint_ , rvalue source arguments can be bound by reference and used after their lifetime ends.

I think this issue is already handled by the `LIFETIMEBOUND` annotation in `GetSource()`.

When I build your test, the compiler shows `warning: temporary bound to local reference '_source' will be destroyed at the end of the full-expression [-Wdangling]` on line `LogDebug(make_source(), "temporary %s", "source");`.

So I think any code written in that unsafe style should be caught locally or by CI, and not be an issue. Your suggestion https://github.com/bitcoin/bitcoin/pull/30343#discussion_r2868278108 to restrict `GetSource()` overloading so it can't bind to lvalues could also make sense if `LIFETIMEBOUND` warning isn't sufficient, but I feel like just using `LIFETIMEBOUND` avoids unnecessary complexity and provides a better diagnostic.

stickies-v committed:
log test: add some test coverage on LogAcceptCategory

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

ryanofsky committed:
log refactor: Drop Entry::should_ratelimit field

Drop Entry::should_ratelimit field in favor of Options::ratelimit field.

Dropping the Entry::should_ratelimit field makes buffered log messages
in m_msgs_before_open smaller, and avoids potential confusion because
the field is stored after rate limiting has been applied, so just gets
ignored. Conceptually it also makes sense to treat the ratelimit option
as an input to the logger rather than as an attribute of the message
being logged.

ryanofsky committed:
log refactor: log macro rewrite

Rewrite log macros to fix a number of issues: unnecessary strprintf
calls during fuzzing, confusing error messages when macros are called
with the wrong arguments, duplicated code with unexplained differences
and undocumented assumptions.

Since this is a rewrite, reading the new code in log.h first should be clearer
than starting from the diff.

Specific benefits of the new implementation are:

- Functionality is implemented once in a single `LOG_EMIT` macro instead of
  multiple times in different macros with diverging code paths and unexplained
  differences.

- Unnecessary `strprintf` calls are now skipped when logging is disabled (in
  bitcoind when `-noprinttoconsole -nodebuglogfile` options are used, and in
  tests and kernel applications when DisableLogging is called). This change
  should not affect bitcoind noticeably, but could speed up fuzz tests calling
  functions which log.

- Clearer error messages: If you pass a category to a macro which does not
  accept it, or forget to pass a category to a macro which requires it, you
  will see a direct message telling you to add or remove the category instead
  of expanded macro syntax errors.

- Previously it was possible to call `detail_LogWithSrcLoc` with
  inconsistent or invalid arguments, for example bypassing ratelimiting
  without it being explicit. New `LOG_EMIT` macro is compile-time safe
  and enforces the exact same restrictions as other macros.

- Previously "always evaluate arguments" behavior at Info and higher levels
  looked accidental and was undocumented
  (https://github.com/bitcoin/bitcoin/pull/34374#discussion_r2734793117). Now
  the behavior is documented and explicit.

- `NO_RATE_LIMIT` special case used to bypass rate limiting is dropped.
  The `LOG_EMIT` macro can control rate limiting (and other options if
  they are added in the future) while still using the same ratelimiting
  defaults as other macros.

Co-Authored-By: Hodlinator <172445034+hodlinator@users.noreply.github.com>
Co-Authored-By: stickies-v <stickies-v@protonmail.com>

ryanofsky committed:
Merge branch 'pr/relog' into pr/bclog

ryanofsky committed:
Merge branch 'pr/bclog' into pr/gwlog

ryanofsky committed:
test-each-commit: Increase fetch depth

Needed due to base PR, can be dropped before merge
